### DISCUSSION TITLE
Notebook Kernel Picker for custom notebook extensions

### DISCUSSION THREAD

**Discussion by ehassaan** [microsoft/vscode-discussions#1694] [Extension Development QnA] [answered]
I am working on an extension that provides custom notebook. Right now I am using NotebookController implementation which appears as a kernel in the kernel picker. But I want this list to be dynamic similar to how Jupyter notebook shows. Is there any way to provide kernel picker options for my notebook or even show the same options as shown in Jupyter extension?

After some research I found the proposed API NotebookKernelProvider that seems to provide the similar functionality. Is there any plan to move this API to stable?
 https://github.com/microsoft/vscode/issues/103883

For reference, this is my extension,
https://marketplace.visualstudio.com/items?itemName=ducklake.ducklab

**Comment by isidorn** [comment]
@rebornix might be able to answer this one.

I do not think it is planned for next 3 months. But sharing your scenario and any feedback on the API helps us finalise it sooner.

**Reply by ehassaan** [reply]
Thanks @isidorn. I published next version of my extension with python support. I had to rely on default python environment picker provided by microsoft/python extension. There are two limitations with my current workaround,

1. If the user clicks kernel picker then it just shows my NotebookControllers. I want to have a custom handler/provider for kernel picker. Kernels have to be created dynamically based on python environment and configuration selected but NotebookController has to be initialized beforehand. So I am just getting the environment from python extension and using default values for other configurations.

2. Right now, I am relying on default python environment picker in bottom ribbon for environment selection. But if user's cursor is in an SQL cell in the notebook then python environment picker disappears. I might be able to do a workaround for this, but it won't be needed if I could just use kernel picker to request python environment selection in the first place.

**Reply by burkeholland** [reply, marked as answer]
@ehassaan Could you convert this to a feature request / issue on the VS Code repo?

https://github.com/microsoft/vscode
